File: 2025/03/main.cpp

#include <cstddef>
#include <fstream>
#include <iostream>
#include <string>
#include <string_view>

namespace aoc_o3 {

int largest2(std::string_view s) {
  size_t i1{}, i2{1};
  char c1{s.at(0)}, c2{s.at(1)};
  for (size_t i{1}; i < s.size(); ++i) {
    if (s.at(i) > c1 and i != s.size() - 1) {
      i1 = i;
      c1 = s.at(i);
      i2 = i + 1;
      c2 = s.at(i + 1);
    } else if (s.at(i) > c2) {
      i2 = i;
      c2 = s.at(i);
    }
  }
  return (c1 - '0') * 10 + (c2 - '0');
}

std::string largestn(std::string_view s, int idx) {
  if (idx < 1) {
    return "";
  }
  size_t len{s.size()};
  if (len < idx) {
    return "";
  }
  char hi{s.at(0)};
  size_t hi_idx{0};
  for (auto i{0}; i <= len - idx; ++i) {
    if (s.at(i) > hi) {
      hi = s.at(i);
      hi_idx = i;
    }
  }
  return hi + largestn(s.substr(hi_idx + 1, len - hi_idx - 1), idx - 1);
}

} // namespace aoc_o3

int main(int argc, char *argv[]) {
  if (argc < 2) {
    std::cerr << "missing input argv" << std::endl;
    return 1;
  }
  std::ifstream input(argv[1]);
  if (!input) {
    std::cerr << "input file cannot be opened" << std::endl;
    return 1;
  }
  std::string line{};
  long long sum{};
  while (std::getline(input, line)) {
    long long n = std::stoll(aoc_o3::largestn(line, 12));
    // std::cout << line << " -> " << n << "\n";
    sum += n;
  }
  std::cout << sum << std::endl;
  return 0;
};

File: 2025/03/main.zig
const std = @import("std");

pub fn largestn(s: []u8, idx: usize, buf: []u8) void {
    if (idx < 1 or s.len < idx) return;
    var hi: u8 = s[0];
    var hi_i: usize = 0;

    for (s[0 .. s.len - idx], 0..) |c, i| {
        if (c > hi) {
            hi = c;
            hi_i = i;
        }
    }
    buf[0] = hi;
    largestn(s[hi_i + 1 ..], idx - 1, buf[1..]);
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const argv = try std.process.argsAlloc(alloc);
    if (argv.len < 2) {
        std.debug.print("missing input argv\n", .{});
        return;
    }
    const file = try std.fs.openFileAbsolute(argv[1], .{ .mode = .read_only });
    defer file.close();

    var buffer: [1024]u8 = undefined;
    var file_reader = file.reader(&buffer);
    const reader = &file_reader.interface;

    var sum: u64 = 0;
    var num_buf: [64]u8 = undefined;
    while (reader.takeDelimiterInclusive('\n')) |line| {
        largestn(line, 12, &num_buf);
        sum += try std.fmt.parseInt(u64, num_buf[0..12], 10);
        // std.debug.print("{s}\n", .{num_buf[0..2]});
    } else |err| {
        if (err != error.EndOfStream) {
            std.debug.print("{}", .{err});
        }
    }
    std.debug.print("{d}\n", .{sum});
}
